common/lib/transport/protocol.ts

```diff
@@ -4,43 +4,65 @@ import EventEmitter from '../util/eventemitter';
 import Logger from '../util/logger';
 import MessageQueue from './messagequeue';
 import ErrorInfo from '../types/errorinfo';
+import Transport from './transport';
+import { ErrCallback } from '../../types/utils';
 
-var Protocol = (function() {
-	var actions = ProtocolMessage.Action;
+const actions = ProtocolMessage.Action;
 
-	function Protocol(transport) {
-		EventEmitter.call(this);
+export class PendingMessage {
+	message: ProtocolMessage;
+	callback?: ErrCallback;
+	merged: boolean;
+	sendAttempted: boolean;
+	ackRequired: boolean;
+
+	constructor(message: ProtocolMessage, callback?: ErrCallback) {
+		this.message = message;
+		this.callback = callback;
+		this.merged = false;
+		const action = message.action;
+		this.sendAttempted = false;
+		this.ackRequired = (action == actions.MESSAGE || action == actions.PRESENCE);
+	}
+}
+
+class Protocol extends EventEmitter {
+	transport: Transport;
+	messageQueue: MessageQueue;
+
+	constructor(transport: Transport) {
+		super();
 		this.transport = transport;
 		this.messageQueue = new MessageQueue();
-		var self = this;
-		transport.on('ack', function(serial, count) { self.onAck(serial, count); });
-		transport.on('nack', function(serial, count, err) { self.onNack(serial, count, err); });
+		transport.on('ack', (serial: number, count: number) => { this.onAck(serial, count); });
+		transport.on('nack', (serial: number, count: number, err: ErrorInfo) => { this.onNack(serial, count, err); });
 	}
-	Utils.inherits(Protocol, EventEmitter);
 
-	Protocol.prototype.onAck = function(serial, count) {
+	static PendingMessage = PendingMessage;
```

Comment:
A good question. @owenpearson ?

Reply:
Good point. This is an artefact of the old way the lib was modularised so that we could export more than one thing from the same file. Since we're using webpack and ES6 modules now this was no longer needed so I've fixed it in 43487e9.